### PR TITLE
functionality for optional variables validations

### PR DESCRIPTION
**Pull request by badri-pathak**
## Pull request checklist
This PR is for enhancement added related to validation of the variables from the optional-config Map "ibm-spectrum-scale-csi-config"

Currently we are supporting these many variable from the configMap:

- VAR_DRIVER_LOGLEVEL
- VAR_DRIVER_PERSISTENT_LOG
- VAR_DRIVER_NODEPUBLISH_METHOD
- DRIVER_UPGRADE_MAXUNAVAILABLE
- VAR_DRIVER_VOLUME_STATS_CAPABILITY

[[CSI 2.10] Validate optional configmap parameters in operator](https://github.ibm.com/ibmspectrumscale/csi-tracker/issues/208)

```release-notes

```

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

- In the current code, there are not much validations available in the code related to key and respective values.

## What is the new behavior?

- These changes will enforce user to provide valid data into optional-config map to get configured where ever applicable like env of driver container and daemonset upgrade strategy. 
 Below are the checks applied in the new code. All the env variable can be view in the driver container/spec. If right value is not provided it won't be accepted by the code, log can be seen to verify. There won't be any stop, if wrong values provided, default values for all these variables will be set. 
 

1. Only five key:value pairs are accepted for the code from the optional configMap.("VAR_DRIVER_LOGLEVEL","VAR_DRIVER_PERSISTENT_LOG","VAR_DRIVER_NODEPUBLISH_METHOD","DRIVER_UPGRADE_MAXUNAVAILABLE","VAR_DRIVER_VOLUME_STATS_CAPABILITY")
2.  Accepted value for VAR_DRIVER_LOGLEVEL is "TRACE", "DEBUG", "INFO", "WARNING", "ERROR", "FATAL".
3. Accepted value for VAR_DRIVER_PERSISTENT_LOG is "ENABLED", "DISABLED"
4. Accepted value for VAR_DRIVER_NODEPUBLISH_METHOD is "SYMLINK", "BINDMOUNT"
5. Accepted value for DRIVER_UPGRADE_MAXUNAVAILABLE is natural no with % symbol. e.g 50%, 40% not 50.1%,50#,50
6. Accepted value for VAR_DRIVER_VOLUME_STATS_CAPABILITY is "ENABLED", "DISABLED"
7. Default values for each are: LOGLEVEL="INFO", PERSISTENT_LOG = "DISABLED", NODEPUBLISH_METHOD = "BINDMOUNT", DRIVER_UPGRADE_MAXUNAVAILABLE =1,VAR_DRIVER_VOLUME_STATS_CAPABILITY="ENABLED"

## How risky is this change?
- [x] Small, isolated change
- [ ] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

**Comment by badri-pathak** [comment]
Images:
quay.io/badri_pathak/ibm-spectrum-scale-csi-operator:variable_validation_v2
quay.io/badri_pathak/ibm-spectrum-scale-csi-driver:variable_validation_v2

File: operator/controllers/csiscaleoperator_controller.go
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:2264]: invalidEnvValue condition is not needed

File: operator/controllers/csiscaleoperator_controller.go
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:2267]: SetStatusAndRaiseEvent is not needed for this info message

File: operator/controllers/csiscaleoperator_controller.go
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:460]: This logger.Error statement doesn't seem right, earlier one was fine.

File: operator/controllers/csiscaleoperator_controller.go
Comment (amdabhad) [reply operator/controllers/csiscaleoperator_controller.go:2320]: Discussed with Badri and even though the defaults are set, keeping invalidEnvValue to print wrong input values for the valid keys.

File: operator/controllers/csiscaleoperator_controller.go
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:449]: We shouldn't validate the configmap data every time in the reconcile loop, validation should be done only when there is a change in the configmap data.

**Comment by badri-pathak** [comment]
Images:
quay.io/badri_pathak/ibm-spectrum-scale-csi-operator:variable_validation_v5
quay.io/badri_pathak/ibm-spectrum-scale-csi-driver:variable_validation_v5

**Review by amdabhad — APPROVED** [review]

**Comment by saurabhwani5** [comment]
Adding issues Here :
1. Log level is not working as per functionality , even if I'm setting it to FATAL I can see all the logs :
```
[root@saurabh9-master functional-tests]# oc logs ibm-spectrum-scale-csi-c78zg
I0526 14:20:03.131807       1 main.go:68] [LOGLEVEL] found in the env : FATAL
I0526 14:20:03.131914       1 main.go:71] [PERSISTENT_LOG] found in the env : DISABLED
I0526 14:20:03.131921       1 main.go:74] [NODEPUBLISH_METHOD] found in the env : BINDMOUNT
I0526 14:20:03.131964       1 main.go:109] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] Version Info: commit ()
I0526 14:20:03.132010       1 gpfs.go:122] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs GetScaleDriver
I0526 14:20:03.132018       1 gpfs.go:198] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs SetupScaleDriver. name: spectrumscale.csi.ibm.com, version: 2.10.0, nodeID: saurabh9-worker-1.fyre.ibm.com
I0526 14:20:03.132040       1 gpfs.go:245] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs PluginInitialize
I0526 14:20:03.132343       1 connectors.go:121] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] connector GetSpectrumScaleConnector
I0526 14:20:03.172018       1 gpfs.go:282] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] IBM Storage Scale: Plugin initialized
I0526 14:20:03.172047       1 gpfs.go:152] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs AddVolumeCapabilityAccessModes
I0526 14:20:03.172055       1 gpfs.go:163] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs AddControllerServiceCapabilities
I0526 14:20:03.172068       1 gpfs.go:231] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] volume stats capabililty is enabled
I0526 14:20:03.172073       1 gpfs.go:174] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs AddNodeServiceCapabilities
I0526 14:20:03.172080       1 gpfs.go:127] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs NewIdentityServer
I0526 14:20:03.172085       1 gpfs.go:145] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs NewNodeServer
I0526 14:20:03.172089       1 gpfs.go:134] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] gpfs NewControllerServer
I0526 14:20:03.172095       1 gpfs.go:287] [181ab5fe-5c45-4ec5-8a9c-865ccaa2645a] Driver: spectrumscale.csi.ibm.com version: 2.10.0
I0526 14:20:03.172197       1 server.go:96] Start listening with scheme unix, addr /var/lib/kubelet/plugins/spectrumscale.csi.ibm.com/csi.sock
I0526 14:20:03.172591       1 server.go:118] Listening for connections on address: &net.UnixAddr{Name:"/var/lib/kubelet/plugins/spectrumscale.csi.ibm.com/csi.sock", Net:"unix"}
I0526 14:20:03.845696       1 utils.go:54] [2f4bd808-f366-4113-9daf-1a442dcebe98] GRPC call: /csi.v1.Identity/GetPluginInfo
I0526 14:20:03.845734       1 identityserver.go:79] [2f4bd808-f366-4113-9daf-1a442dcebe98] Using default GetPluginInfo
I0526 14:20:03.845742       1 utils.go:61] [2f4bd808-f366-4113-9daf-1a442dcebe98] GRPC response: name:"spectrumscale.csi.ibm.com" vendor_version:"2.10.0"
I0526 14:20:03.845867       1 utils.go:65] [2f4bd808-f366-4113-9daf-1a442dcebe98] Time taken to execute GRPC request(in milli): 0
I0526 14:20:04.536980       1 utils.go:54] [f4346120-5f20-4ce3-aef0-8825cd705e36] GRPC call: /csi.v1.Identity/GetPluginInfo
I0526 14:20:04.537007       1 identityserver.go:79] [f4346120-5f20-4ce3-aef0-8825cd705e36] Using default GetPluginInfo
I0526 14:20:04.537016       1 utils.go:61] [f4346120-5f20-4ce3-aef0-8825cd705e36] GRPC response: name:"spectrumscale.csi.ibm.com" vendor_version:"2.10.0"
I0526 14:20:04.537090       1 utils.go:65] [f4346120-5f20-4ce3-aef0-8825cd705e36] Time taken to execute GRPC request(in milli): 0
I0526 14:20:04.794418       1 utils.go:54] [74dc001e-1039-4df0-8c45-96cda1f3909b] GRPC call: /csi.v1.Node/NodeGetInfo
I0526 14:20:04.794460       1 utils.go:61] [74dc001e-1039-4df0-8c45-96cda1f3909b] GRPC response: node_id:"saurabh9-worker-1.fyre.ibm.com"
I0526 14:20:04.794633       1 utils.go:65] [74dc001e-1039-4df0-8c45-96cda1f3909b] Time taken to execute GRPC request(in milli): 0
I0526 14:20:55.343221       1 utils.go:54] [dca3f9c0-a955-4f3b-b776-ef5e590849d9] GRPC call: /csi.v1.Identity/Probe
E0526 14:20:55.371598       1 identityserver.go:62] [dca3f9c0-a955-4f3b-b776-ef5e590849d9] Probe: GPFS component on node saurabh9-worker-1.fyre.ibm.com is not healthy. Error: <nil>
I0526 14:20:55.371628       1 utils.go:61] [dca3f9c0-a955-4f3b-b776-ef5e590849d9] GRPC response: ready:<value:true >
I0526 14:20:55.371829       1 utils.go:65] [dca3f9c0-a955-4f3b-b776-ef5e590849d9] Time taken to execute GRPC request(in milli): 28
I0526 14:20:55.415672       1 utils.go:54] [6236d6ae-1cf5-43ec-91b9-c7814703b2d6] GRPC call: /csi.v1.Identity/GetPluginInfo
I0526 14:20:55.415715       1 identityserver.go:79] [6236d6ae-1cf5-43ec-91b9-c7814703b2d6] Using default GetPluginInfo
I0526 14:20:55.415724       1 utils.go:61] [6236d6ae-1cf5-43ec-91b9-c7814703b2d6] GRPC response: name:"spectrumscale.csi.ibm.com" vendor_version:"2.10.0"
I0526 14:20:55.415837       1 utils.go:65] [6236d6ae-1cf5-43ec-91b9-c7814703b2d6] Time taken to execute GRPC request(in milli): 0
I0526 14:20:55.416770       1 utils.go:54] [c45caf16-3fcd-48a7-bc61-ece8a36e1cee] GRPC call: /csi.v1.Identity/GetPluginCapabilities
I0526 14:20:55.416807       1 utils.go:61] [c45caf16-3fcd-48a7-bc61-ece8a36e1cee] GRPC response: capabilities:<service:<type:CONTROLLER_SERVICE > >
I0526 14:20:55.417166       1 utils.go:65] [c45caf16-3fcd-48a7-bc61-ece8a36e1cee] Time taken to execute GRPC request(in milli): 1
I0526 14:20:55.418011       1 utils.go:54] [f33d5317-d853-4c1b-84a5-da4f898c5eec] GRPC call: /csi.v1.Controller/ControllerGetCapabilities
I0526 14:20:55.418046       1 controllerserver.go:1686] [f33d5317-d853-4c1b-84a5-da4f898c5eec] ControllerGetCapabilities called with req: &csi.ControllerGetCapabilitiesRequest{XXX_NoUnkeyedLiteral:struct {}{}, XXX_unrecognized:[]uint8(nil), XXX_sizecache:0}
I0526 14:20:55.418097       1 utils.go:61] [f33d5317-d853-4c1b-84a5-da4f898c5eec] GRPC response: capabilities:<rpc:<type:CREATE_DELETE_VOLUME > > capabilities:<rpc:<type:PUBLISH_UNPUBLISH_VOLUME > > capabilities:<rpc:<type:CREATE_DELETE_SNAPSHOT > > capabilities:<rpc:<type:EXPAND_VOLUME > > capabilities:<rpc:<type:CLONE_VOLUME > >
I0526 14:20:55.418377       1 utils.go:65] [f33d5317-d853-4c1b-84a5-da4f898c5eec] Time taken to execute GRPC request(in milli): 0
I0526 14:22:02.724890       1 utils.go:54] [e0f69a4f-dc8b-4320-85c4-3feab7884955] GRPC call: /csi.v1.Identity/Probe
E0526 14:22:02.755960       1 identityserver.go:62] [e0f69a4f-dc8b-4320-85c4-3feab7884955] Probe: GPFS component on node saurabh9-worker-1.fyre.ibm.com is not healthy. Error: <nil>
I0526 14:22:02.756014       1 utils.go:61] [e0f69a4f-dc8b-4320-85c4-3feab7884955] GRPC response: ready:<value:true >
I0526 14:22:02.756104       1 utils.go:65] [e0f69a4f-dc8b-4320-85c4-3feab7884955] Time taken to execute GRPC request(in milli): 32
I0526 14:23:17.397789       1 utils.go:54] [d7120185-bf97-4854-aba6-3292a899fdd3] GRPC call: /csi.v1.Controller/DeleteVolume
I0526 14:23:17.397867       1 controllerserver.go:1565] [d7120185-bf97-4854-aba6-3292a899fdd3] DeleteVolume [volume_id:"0;0;11840458782091175455;9F730B0A:6432E6FC;;;/ibm/fs1/spectrum-scale-csi-volume-store/.volumes/pvc-bcb68ae6-a34a-4dbe-ae9b-5c1edb5a2aba" ]
I0526 14:23:17.398020       1 gpfs.go:185] [d7120185-bf97-4854-aba6-3292a899fdd3] gpfs ValidateControllerServiceRequest
I0526 14:23:17.399552       1 utils.go:54] [33248b72-7395-42dd-8082-15db32226322] GRPC call: /csi.v1.Controller/CreateVolume
I0526 14:23:17.399630       1 controllerserver.go:574] [33248b72-7395-42dd-8082-15db32226322] create volume req: name:"pvc-8599f187-6711-4a1e-9c65-a17877e90373" capacity_range:<required_bytes:1073741824 > volume_capabilities:<mount:<fs_type:"gpfs" > access_mode:<mode:MULTI_NODE_MULTI_WRITER > > parameters:<key:"clusterId" value:"11840458782091175455" > parameters:<key:"csi.storage.k8s.io/pv/name" value:"pvc-8599f187-6711-4a1e-9c65-a17877e90373" > parameters:<key:"csi.storage.k8s.io/pvc/name" value:"pvc-elegant-fox-893" > parameters:<key:"csi.storage.k8s.io/pvc/namespace" value:"ibm-spectrum-scale-csi-test" > parameters:<key:"volBackendFs" value:"fs1" >
I0526 14:23:17.399826       1 gpfs.go:185] [33248b72-7395-42dd-8082-15db32226322] gpfs ValidateControllerServiceRequest
I0526 14:23:17.399910       1 controllerserver.go:527] [33248b72-7395-42dd-8082-15db32226322] getPrimaryClusterDetails
I0526 14:23:17.435100       1 controllerserver.go:551] [33248b72-7395-42dd-8082-15db32226322] symlinkDirPath [/ibm/fs1/spectrum-scale-csi-volume-store/.volumes], symlinkDirRelPath [spectrum-scale-csi-volume-store/.volumes]
I0526 14:23:17.435169       1 controllerserver.go:674] [33248b72-7395-42dd-8082-15db32226322] volume:[pvc-8599f187-6711-4a1e-9c65-a17877e90373] - check if primary fileset [spectrum-scale-csi-volume-store] is linked

```

**Comment by badri-pathak** [comment]
@hemalathagajendran can we check this issue together?

**Review by Jainbrt — CHANGES_REQUESTED** [review]
have we not considered VAR_DRIVER_VOLUME_STATS_CAPABILITY variable for valid values ?

**Comment by badri-pathak** [comment]
> have we not considered VAR_DRIVER_VOLUME_STATS_CAPABILITY variable for valid values ?

Added VAR_DRIVER_VOLUME_STATS_CAPABILITY in the latest commit. Default will be enabled. Please valid once.

**Comment by badri-pathak** [comment]
Latest images: 
quay.io/badri_pathak/ibm-spectrum-scale-csi-driver:variable_validation_v10
quay.io/badri_pathak/ibm-spectrum-scale-csi-operator:variable_validation_v10

**Review by saurabhwani5 — CHANGES_REQUESTED** [review]
`DRIVER_UPGRADE_MAXUNAVAILABLE` is not setting the value

**Comment by amdabhad** [comment]
@Jainbrt, @saurabhwani5, events will be added in another PR and logger issue present in dev will be fixed under https://github.com/IBM/ibm-spectrum-scale-csi/issues/979, please approve the PR if there is nothing else expected.

**Review by Jainbrt — DISMISSED** [review]

**Comment by saurabhwani5** [comment]
I have  following Images for final review, Thanks @badri-pathak for all the required changes:
```
quay.io/badri_pathak/ibm-spectrum-scale-csi-operator:variable_validation_v23
quay.io/badri_pathak/ibm-spectrum-scale-csi-driver:variable_validation_v16
```

**Review by saurabhwani5 — APPROVED** [review]

**Review by Jainbrt — DISMISSED** [review]

File: operator/controllers/syncer/csi_node.go
Comment (amdabhad) [inline operator/controllers/syncer/csi_node.go:118]: This log statement seems unnecessary as cmEnvVars will never have upgrade strategy due to condition at line no 105.

**Review by amdabhad — APPROVED** [review]

**Pull request merged**
